File: ThreatModelDfdService/Services/Impl/DataFlowService.cs
using ThreatModelDfdService.Data.DTO;
using ThreatModelDfdService.Model.Context;
using ThreatModelDfdService.Model.Entity;
using ThreatModelDfdService.Repositories;

namespace ThreatModelDfdService.Services.Impl;

public class DataFlowService(
    DfdElementService dfdElementService,
    IRepository<DataFlow> dataFlowRepository,
    MSSQLContext context
)
{
    public async Task CreateOrUpdateDataFlow(DataFlowRequestDTO dto)
    {
        long? sourceElementId = null;
        long? targetElementId = null;
        if(dto.SourceElementIdentifier != null)
        {
            DfdElement sourceElement = dfdElementService.GetByUUID(dto.SourceElementIdentifier);
            sourceElementId = sourceElement.Id;
        }
        if(dto.TargetElementIdentifier != null)
        {
            DfdElement targetElement = dfdElementService.GetByUUID(dto.TargetElementIdentifier);
            targetElementId = targetElement.Id;
        }

        if(dto.Id.HasValue && dto.Id.Value > 0)
        {
            DataFlow dbDataFlow = FindById(dto.Id.Value);
            dbDataFlow.Name = dto.Name;
            dbDataFlow.Description = dto.Description;
            if(sourceElementId.HasValue) dbDataFlow.SourceElementId = sourceElementId;
            if(targetElementId.HasValue) dbDataFlow.TargetElementId = targetElementId;
            dbDataFlow.SourcePosition = dto.SourcePosition;
            dbDataFlow.TargetPosition = dto.TargetPosition;

            await dataFlowRepository.UpdateAsync(dbDataFlow);
        }
        else
        {
            DataFlow newDataFlow = new DataFlow
            {
                Name = dto.Name,
                Description = dto.Description,
                SourceElementId = sourceElementId,
                TargetElementId = targetElementId,
                SourcePosition = dto.SourcePosition,
                TargetPosition = dto.TargetPosition
            };
            await dataFlowRepository.CreateAsync(newDataFlow);
        }
    }

    public List<DataFlowResponseDTO> GetDataFlowsByDfdId(long dfdId)
    {
        List<DataFlow> dataFlows = context.DataFlows.Where(df => df.SourceElement.DfdId == dfdId || df.TargetElement.DfdId == dfdId).ToList();
        return dataFlows.Select(df => new DataFlowResponseDTO(
            df.Id,
            df.Name,
            df.Description,
            df.SourceElement != null ? df.SourceElement.UuidIdentifier : null,
            df.TargetElement != null ? df.TargetElement.UuidIdentifier : null,
            df.SourcePosition,
            df.TargetPosition
        )).ToList();
    }

    public DataFlow FindById(long id)
    {
        DataFlow dbDataFlow = context.DataFlows.FirstOrDefault(df => df.Id == id);
        if (dbDataFlow == null)
        {
            throw new ArgumentException("Data flow with the provided ID does not exist. | Id: " + id);
        }
        return dbDataFlow;
    }
}

File: dfd-service/src/main/java/com/tcc/dfd_service/service/DataFlowService.java
package com.tcc.dfd_service.service;

import com.tcc.dfd_service.dto.DataFlowRequestDTO;
import com.tcc.dfd_service.dto.DataFlowResponseDTO;
import com.tcc.dfd_service.entity.DataFlow;
import com.tcc.dfd_service.entity.Dfd;
import com.tcc.dfd_service.entity.DfdElement;
import com.tcc.dfd_service.repository.DataFlowRepository;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class DataFlowService {

    private final DfdElementService dfdElementService;
    private final DataFlowRepository dataFlowRepository;

    @Transactional
    public void createOrUpdateDataFlow(DataFlowRequestDTO dto, Long dfdId) {
        DfdElement sourceElement = null;
        DfdElement targetElement = null;

        if (dto.sourceElementIdentifier() != null) {
            sourceElement = dfdElementService.getByUUID(dto.sourceElementIdentifier());
        }

        if (dto.targetElementIdentifier() != null) {
            targetElement = dfdElementService.getByUUID(dto.targetElementIdentifier());
        }

        if (dto.id() != null && dto.id() > 0) {
            DataFlow dbDataFlow = findById(dto.id());

            dbDataFlow.setName(dto.name());
            dbDataFlow.setDescription(dto.description());

            if (sourceElement != null) {
                dbDataFlow.setSourceElement(sourceElement);
            }

            if (targetElement != null) {
                dbDataFlow.setTargetElement(targetElement);
            }

            dbDataFlow.setSourcePosition(dto.sourcePosition());
            dbDataFlow.setTargetPosition(dto.targetPosition());

        } else {

            DataFlow newDataFlow = new DataFlow();

            newDataFlow.setName(dto.name());
            newDataFlow.setDescription(dto.description());
            newDataFlow.setSourceElement(sourceElement);
            newDataFlow.setTargetElement(targetElement);
            newDataFlow.setSourcePosition(dto.sourcePosition());
            newDataFlow.setTargetPosition(dto.targetPosition());
            newDataFlow.setDfdId(dfdId);

            dataFlowRepository.save(newDataFlow);
        }
    }

    public List<DataFlowResponseDTO> getDataFlowsByDfdId(Long dfdId) {
        return dataFlowRepository
                .findByDfdId(dfdId)
                .stream()
                .map(df -> new DataFlowResponseDTO(
                        df.getId(),
                        df.getName(),
                        df.getDescription(),
                        df.getSourceElement() != null
                                ? df.getSourceElement().getUuidIdentifier()
                                : null,
                        df.getTargetElement() != null
                                ? df.getTargetElement().getUuidIdentifier()
                                : null,
                        df.getSourcePosition(),
                        df.getTargetPosition()
                ))
                .toList();
    }

    public DataFlow findById(Long id) {
        return dataFlowRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException(
                        "Data flow with the provided ID does not exist. Id: " + id)
                );
    }
}